lxd/certificates.go

```diff
@@ -718,19 +718,58 @@ func doCertificateUpdate(d *Daemon, dbInfo db.Certificate, req api.CertificatePu
 		}
 
 		// Convert to the database type.
-		cert := db.Certificate{
-			// Read-only fields.
+		dbCert := db.Certificate{
 			Certificate: dbInfo.Certificate,
 			Fingerprint: dbInfo.Fingerprint,
+			Restricted:  req.Restricted,
+			Projects:    req.Projects,
+			Name:        req.Name,
+			Type:        reqDBType,
+		}
+
+		if req.Certificate != "" && dbInfo.Certificate != req.Certificate {
```

Comment:
So we need to be very very careful here. The user should be allowed to replace their certificate but only that.

They should not be allowed to change any of the other properties (restricted, projects, name) as at least the first two would cause a massive security issue.

So in the non-admin+TLS block we need to also ensure that the only thing changed is the certificate itself.
And for good measure I'd probably have that codepath reset dbCert to be entirely based on dbInfo instead of req so that should we introduce something new down the line, we don't end up with a security issue.

Reply:
I would also ask for future readers that the logic here be thoroughly commented to explain both the security implications of making changes and exactly the decisions and precautions being used so its clear for future readers/modifications. Also adding tests for trying to bypass the security restrictions would be important too to catch regressions.

Reply:
> Also adding tests for trying to bypass the security restrictions would be important too to catch regressions.

We don't have rbac tests (or at least I couldn't find any) so I believe we cannot test for this.

Reply:
We should add something to `test-lxd-rbac` in lxc-ci.

Reply:
Maybe I misunderstood the code changes here, but is this actually tied to RBAC? the only thing I can see is where its checking that the request isn't from an RBAC admin user? Does this means a normal non-RBAC client could update their certificate? If so presumably we could add tests to the normal test suite for this.

Reply:
> Does this means a normal non-RBAC client could update their certificate?

Yes, that's correct. Clients are always admins.

Reply:
Ah I see, so `rbac.UserIsAdmin()` actually returns true for non-rbac admin users too, which is little confusing IMHO. Perhaps later (not part of this PR) we could clarify the difference between LXD roles and our external RBAC integration, as the two seem a little conflated ATM.

But back to the point in hand, what do you make of this:

https://github.com/lxc/lxd/blob/8dc4aed105b3e9cff253a3a042e1faadde52bf23/lxd/daemon.go#L531-L558

It seems to suggest that we can have non-admin users that are not connected to RBAC authentication, if they are assigned to a restricted project.

https://github.com/lxc/lxd/blob/8dc4aed105b3e9cff253a3a042e1faadde52bf23/lxd/daemon.go#L541

That seems like an approach we could use for testing self-cert renewal?

Reply:
I didn't see or know that. I just tested it and indeed it won't let me change the certificate if I'm restricted.

Reply:
Ah yeah, so we could use that. Someone using a restricted certificate should be allowed to update their own certificate (but not alter the restricted/projects properties) but then shouldn't be allowed to modify other unrelated certificates.